### ISSUE TITLE
Secure media

### ISSUE THREAD

**Issue by perrygeo** [Ecotrust/madrona#417]
This is IMO a huge problem with Django and it sucks that we need to develop a workaround but...

Django writes FileFields to the MEDIA_ROOT dir. As a result, cached files and user-uploaded files gets stored relative to MEDIA. 

But its also where static media is stored so it is wide-open web accessible through apache. 

We need a system whereby user-uploaded and cached content can be hidden behind django-level authentication with sharing permissions, etc

One proposed solution:
a) have a private media dir (MEDIA_ROOT/private/) which we lock down via apache settings and in the urlconf

b) all sensitive FileFields write to 'private/appname/____' 

c) set up an urlpattern to capture private media urls and a session key like:

   /appname/getsomefile/array/1

d) set up a view to authenticate, determine permissions and serve the file back
 (see http://djangosnippets.org/snippets/365/)

Basically a ton of scaffolding work to create a django proxy to bypass apache.

Another option would be to:

a) structure filefields to store data according to a convention whereby we could determine the model, fieldname and pk from the file path:

   /spacing/array/1/cached_spacing_kmz/_____

b) Use django as an apache authentication provider whereby permissions for the object could be determined by parsing the path (eg From the example above, look for Array.objects.get(pk=1) and check with the can_user_view() sharing utility)

http://code.google.com/p/modwsgi/wiki/AccessControlMechanisms

Both of these options suck so I'm hoping someone has a better idea!

Without exception

Google Code Info:
Issue #: 422
Author: perrygeo...@gmail.com
Created On: 2010-08-20T21:52:10.000Z
Closed On:
